org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPEclipseUtils.java

```diff
@@ -397,23 +397,28 @@ public static void applyEdits(IDocument document, List<? extends TextEdit> edits
 			return;
 		}
 
-		IDocumentUndoManager manager = DocumentUndoManagerRegistry.getDocumentUndoManager(document);
-		if (manager != null) {
-			manager.beginCompoundChange();
-		}
-
 		MultiTextEdit edit = new MultiTextEdit();
 		for (TextEdit textEdit : edits) {
 			if (textEdit != null) {
 				try {
 					int offset = LSPEclipseUtils.toOffset(textEdit.getRange().getStart(), document);
 					int length = LSPEclipseUtils.toOffset(textEdit.getRange().getEnd(), document) - offset;
+					if (length < 0) {
+ 						// Must be a bad location: we bail out to avoid corrupting the document.
+ 						LanguageServerPlugin.logInfo("Invalid location information found during formatting"); //$NON-NLS-1$
+ 						return;
```

Comment:
maybe just 
throw new BadLocationException("Negative length during formatting"); //$NON-NLS-1$
?

Reply:
I don't think it makes any difference at the moment, since this method works atomically and none of the edits will get applied if one of them causes an error. None of the current clients of this method would behave differently, but maybe throwing an exception would be more future-proof, as it might allow future consumers of it to bail out of other actions if it flagged up an inconsistency between client and server?

Reply:
Yes, I agree it is not very important. I think it would be nicer code, but probably just a matter of personal preference.